### PR TITLE
stabilize member constraints

### PR DESCRIPTION
**Pull request by nikomatsakis**
Stabilizes the use of "member constraints" in solving `impl Trait` bindings. This is a step towards stabilizing a "MVP" of "named impl Trait".

# Member constraint stabilization report

| Info | |
| --- | --- |
| Tracking issue | [rust-lang/rust#61997](https://github.com/rust-lang/rust/issues/61997) |
| Implementation history | [rust-lang/rust#61775] |
| rustc-dev-guide coverage | [link](https://rustc-dev-guide.rust-lang.org/borrow_check/region_inference/member_constraints.html) |
| Complications | [rust-lang/rust#61773] |

[rust-lang/rust#61775]: https://github.com/rust-lang/rust/pull/61775
[rust-lang/rust#61773]: https://github.com/rust-lang/rust/issues/61773

## Background

Member constraints are an extension to our region solver that was introduced to make async fn region solving tractable. There are used in situations like the following:

```rust
fn foo<'a, 'b>(...) -> impl Trait<'a, 'b> { .. }
```

The problem here is that every region R in the hidden type must be equal to *either* `'a` *or* `'b` (or `'static`). This cannot be expressed simply via 'outlives constriants' like `R: 'a`. Therefore, we introduce a 'member constraint' `R member of ['a, 'b]`.

These constraints were introduced in [rust-lang/rust#61775]. At the time, we kept them feature gated and used them only for `impl Trait` return types that are derived from `async fn`. The intention, however, was always to support them in other contexts once we had time to gain more experience with them. 

**In the time since their introduction, we have encountered no surprises or bugs due to these member constraints.** They are tested extensively as part of every async function that involves multiple unrelated lifetimes in its arguments.

## Tests

The behavior of member constraints is covered by the following tests:

* [`src/test/ui/async-await/multiple-lifetimes`](https://github.com/rust-lang/rust/tree/20e032e65007ff1376e8480c1fbdb0a5068028fa/src/test/ui/async-await/multiple-lifetimes) -- tests using the async await, which are mostly already stabilized
* [`src/test/ui/impl-trait/multiple-lifetimes.rs`](https://github.com/rust-lang/rust/blob/20e032e65007ff1376e8480c1fbdb0a5068028fa/src/test/ui/impl-trait/multiple-lifetimes.rs)
* [`src/test/ui/impl-trait/multiple-lifetimes/ordinary-bounds-unsuited.rs`](https://github.com/rust-lang/rust/blob/20e032e65007ff1376e8480c1fbdb0a5068028fa/src/test/ui/impl-trait/multiple-lifetimes/ordinary-bounds-unsuited.rs)
* [`src/test/ui/async-await/multiple-lifetimes/ret-impl-trait-fg.rs`](https://github.com/rust-lang/rust/blob/20e032e65007ff1376e8480c1fbdb0a5068028fa/src/test/ui/async-await/multiple-lifetimes/ret-impl-trait-fg.rs)
* [`src/test/ui/async-await/multiple-lifetimes/ret-impl-trait-one.rs`](https://github.com/rust-lang/rust/blob/20e032e65007ff1376e8480c1fbdb0a5068028fa/src/test/ui/async-await/multiple-lifetimes/ret-impl-trait-one.rs)

These tests cover a number of scenarios:

* `-> implTrait<'a, 'b>` with unrelated lifetimes `'a` and `'b`, as described above
* `async fn` that returns an `impl Trait` like the previous case, which desugars to a kind of "nested" impl trait like `impl Future<Output = impl Trait<'a, 'b>>`

## Potential concerns

There is a potential interaction with `impl Trait` on local variables, described in [rust-lang/rust#61773]. The challenge is that if you have a program like:

```rust=
trait Foo<'_> { }
impl Foo<'_> for &u32 { }

fn bar() {
  let x: impl Foo<'_> = &44; // let's call the region variable for `'_` `'1`
}
```

then we would wind up with `'0 member of ['1, 'static]`, where `'0` is the region variable in the hidden type (`&'0 u32`) and `'1` is the region variable in the bounds `Foo<'1>`. This is tricky because both `'0` and `'1` are being inferred -- so making them equal may have other repercussions.

That said, `impl Trait` in bindings are not stable, and the implementation is pretty far from stabilization. Moreover, the difficulty highlighted here is not due to the presence of member constraints -- it's inherent to the design of the language. In other words, stabilizing member constraints does not actually cause us to accept anything that would make this problem any harder. 

So I don't see this as a blocker to stabilization of member constraints; it is potentially a blocker to stablization of `impl trait` in let bindings.

### CODE REVIEW

**Comment by nikomatsakis** [comment]
I've marked this for @rust-lang/lang, but I'd like to cc @rust-lang/compiler, @rust-lang/wg-traits, and especially @matthewjasper for comment.

File: compiler/rustc_feature/src/accepted.rs [hunk 1]
```diff
@@ -281,6 +281,10 @@ declare_features! (
     (accepted, or_patterns, "1.53.0", Some(54883), None),
     /// Allows defining identifiers beyond ASCII.
     (accepted, non_ascii_idents, "1.53.0", Some(55467), None),
+    /// Allows defining identifiers beyond ASCII.
+    (accepted, non_ascii_idents, "1.53.0", Some(55467), None),
```
Comment (jackh726) [inline compiler/rustc_feature/src/accepted.rs:285]: This is duplicated.

File: src/doc/unstable-book/src/language-features/member-constraints.md [hunk 2]
```diff
@@ -1,29 +0,0 @@
-# `member_constraints`
```
Comment (jackh726) [inline src/doc/unstable-book/src/language-features/member-constraints.md:1]: Is this worth adding to the reference? Not sure where it would go though.

**Review by jackh726 — APPROVED** [review]
Small typo.

But overall, this seems extremely straightforward to me. From a language design standpoint, I don't see why the lifetimes in an `impl Trait` *should have to relate to each other* (we don't have that requirement for structs, for example). From a implementation standpoint, the fact that this has been effectively stable in `async fn`s provides good evidence the correct implementation is as correct as it can be and that this is *worth* stabilizing more generally.

File: src/doc/unstable-book/src/language-features/member-constraints.md (hunk 2 above)
Comment (nikomatsakis) [reply src/doc/unstable-book/src/language-features/member-constraints.md:1]: I considered this, but I don't think so. I think the reference should eventually be extended with a more thorough description of the formal type system, but this is really an implementation detail at present. There is good coverage in the rustc-dev-guide.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

As discussed in the lang-team meeting.

File: compiler/rustc_feature/src/accepted.rs [hunk 3]
```diff
@@ -281,6 +281,8 @@ declare_features! (
     (accepted, or_patterns, "1.53.0", Some(54883), None),
     /// Allows defining identifiers beyond ASCII.
     (accepted, non_ascii_idents, "1.53.0", Some(55467), None),
+    /// Allows `impl Trait` with multiple unrelated lifetimes.
+    (accepted, member_constraints, "1.53.0", Some(61997), None),
```
Comment (jackh726) [inline compiler/rustc_feature/src/accepted.rs:285]: I think we're 1.54.0 now.

**Review by jackh726 — APPROVED** [review]
Stabilization PR here is pretty simple (as it should be). One comment, but r=me on this when FCP finishes

File: compiler/rustc_feature/src/accepted.rs (hunk 3 above)
Comment (nikomatsakis) [reply compiler/rustc_feature/src/accepted.rs:285]:
```suggestion
    (accepted, member_constraints, "1.54.0", Some(61997), None),
```

**Comment by lqd** [comment]
(the `compare_mode=nll` tests need blessing)

**Pull request merged**